### PR TITLE
Revert "Revert "Implemented series period detect method using pocketfft lib""

### PR DESCRIPTION
**Pull request by bhavnajindal**
Reverts: [57536](https://github.com/ClickHouse/ClickHouse/pull/57536)
Origin:  [56171](https://github.com/ClickHouse/ClickHouse/pull/56171)

`+ ` Added a fix to handle non-constant arguments by handling `input_rows_count` and returning one output for each row.

`+` Method returns NAN when number of data points are less than four. This  conforms to returning single output corresponding to each input row.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by jkartseva** [comment]
I believe this is missing the revert commit.

**Comment by jkartseva** [comment]
Please also add tests where a function is passed as a parameter, e.g. `seriesPeriodDetectFFT(arrayMap(x -> sin(x / 10), range(1000)))`

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args{{"time_series", &isArray<IDataType>, nullptr, "Array"}};
+        validateFunctionArgumentTypes(*this, arguments, args);
+
+        return std::make_shared<DataTypeFloat64>();
+    }
+
+    ColumnPtr executeImplDryRun(const ColumnsWithTypeAndName &, const DataTypePtr &, size_t input_rows_count) const override
```
Comment (alexey-milovidov) [inline src/Functions/seriesPeriodDetectFFT.cpp:66]: This is not the correct way to solve the problem.

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+        period = std::round(1 / freq);
+        return true;
+    }
+};
+
+REGISTER_FUNCTION(SeriesPeriodDetectFFT)
+{
+    factory.registerFunction<FunctionSeriesPeriodDetectFFT>(FunctionDocumentation{
+        .description = R"(
```
Comment (alexey-milovidov) [inline src/Functions/seriesPeriodDetectFFT.cpp:164]: The description should contain full docs; you can even copy-paste them from the markdown file.

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        std::vector<double> xfreq(spec_len);
+        double step = 0.5 / (spec_len - 1);
+        for (size_t i = 0; i < spec_len; ++i)
+            xfreq[i] = i * step;
+
+        auto freq = xfreq[idx];
+
+        period = std::round(1 / freq);
```
Comment (alexey-milovidov) [inline src/Functions/seriesPeriodDetectFFT.cpp:156]: Why rounding?

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+/*Detect Period in time series data using FFT.
+ * FFT - Fast Fourier transform (https://en.wikipedia.org/wiki/Fast_Fourier_transform)
+ * 1. Convert time series data to frequency domain using FFT.
+ * 2. Remove the 0th(the Dc component) and n/2th the Nyquist frequency
+ * 3. Find the peak value (highest) for dominant frequency component.
+ * 4. Inverse of the dominant frequency component is the period.
+*/
+
+class FunctionSeriesPeriodDetectFFT : public IFunction
```
Comment (alexey-milovidov) [inline src/Functions/seriesPeriodDetectFFT.cpp:43]: Needs performance evaluation.

File: src/Functions/seriesPeriodDetectFFT.cpp (hunk 3 above)
Comment (bhavnajindal) [reply src/Functions/seriesPeriodDetectFFT.cpp:156]: The cycle/period represents number of bins after which the data tends to exhibit repetition. This same period could be later used in time series analysis methods like Series Decompose. These algorithms typically require the period to be a whole number( or to be of type size_t). Therefore, it seemed like a good practical choice to round it off.

File: src/Functions/seriesPeriodDetectFFT.cpp (hunk 4 above)
Comment (bhavnajindal) [reply src/Functions/seriesPeriodDetectFFT.cpp:43]: Ran some tests with large numbers : https://pastila.nl/?00e07f72/2751cbfa358751a9d5f4d8485dd3c392#Y6yuvEIwVq4/Dm+0hhpgug==

Please let me know if we need performance evaluation against some specific benchmarks etc.

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 1f58b0f150b005dc4902bc0d2a8956da2c86de6e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57574/1f58b0f150b005dc4902bc0d2a8956da2c86de6e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7143168923">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57574/1f58b0f150b005dc4902bc0d2a8956da2c86de6e/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: docs/en/sql-reference/functions/time-series-functions.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+- Returns NAN when number of data points are less than four.
+
+Type: [Float64](../../sql-reference/data-types/float.md).
+
+**Examples**
+
+Query:
+
+``` sql
+SELECT seriesPeriodDetectFFT([1, 4, 6, 1, 4, 6, 1, 4, 6, 1, 4, 6, 1, 4, 6, 1, 4, 6, 1, 4, 6]) AS print_0;
```
Comment (jkartseva) [inline docs/en/sql-reference/functions/time-series-functions.md:39]: Could you please add examples with periodic functions?

E.g. sinusoid or https://en.wikipedia.org/wiki/Triangle_wave:

```
SELECT seriesPeriodDetectFFT(arrayMap(x -> (abs(((x - 1) % 4) - 2) - 1), range(1000)))

Query id: 6911842f-1d14-4688-9b4f-adfc79f5ea19

┌─seriesPeriodDetectFFT(arrayMap(lambda(tuple(x), minus(abs(minus(modulo(minus(x, 1), 4), 2)), 1)), range(1000)))─┐
│                                                                                                               4 │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

```

Side question: How would it perform on asymptotic functions (e.g. `tan(x)`)?

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        const ColumnVector<T> * src_data_concrete = checkAndGetColumn<ColumnVector<T>>(&src_data);
+        if (!src_data_concrete)
+            return false;
+
+        const PaddedPODArray<T> & src_vec = src_data_concrete->getData();
+
+        size_t len = src_vec.size();
+        if (len < 4)
+        {
+            period = NAN; // Atleast four data points are required to detect period
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetectFFT.cpp:108]: Spellcheck.

Comment (jkartseva) [reply src/Functions/seriesPeriodDetectFFT.cpp:108]: IMO the previous version with the `ErrorCodes::BAD_ARGUMENTS` exception was more explicit than returning 'nan'.

File: docs/en/sql-reference/functions/time-series-functions.md (hunk 5 above)
Comment (bhavnajindal) [reply docs/en/sql-reference/functions/time-series-functions.md:39]: Added more test-cases (with triangle and square waves) and update examples too. I also compared the method with python for asymptotic functions: https://pastila.nl/?028974ec/f182f872d708823af2d1b832180ff663#krxGEW8PRVF+I5R5/N+uwQ==
The results are similar. Also, FFT is suited towards detecting period in a time series data.

File: src/Functions/seriesPeriodDetectFFT.cpp (hunk 6 above)
Comment (bhavnajindal) [reply src/Functions/seriesPeriodDetectFFT.cpp]: yes, the exception was better in terms of explicit. But, it became imperative for me to change this to conform with CH design of returning exactly one value for every input row.

File: tests/queries/0_stateless/02813_series_period_detect.sql [hunk 7]
```diff
@@ -0,0 +1,12 @@
+-- Tags: no-fasttest
+
+SELECT seriesPeriodDetectFFT([139, 87, 110, 68, 54, 50, 51, 53, 133, 86, 141, 97, 156, 94, 149, 95, 140, 77, 61, 50, 54, 47, 133, 72, 152, 94, 148, 105, 162, 101, 160, 87, 63, 53, 55, 54, 151, 103, 189, 108, 183, 113, 175, 113, 178, 90, 71, 62, 62, 65, 165, 109, 181, 115, 182, 121, 178, 114, 170]);
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply tests/queries/0_stateless/02813_series_period_detect.sql:3]: Add a test with more than one row.

[the remark this answers is not in the document]
Comment (bhavnajindal) [reply tests/queries/0_stateless/02813_series_period_detect.sql:8]: Added

File: tests/queries/0_stateless/02813_series_period_detect.sql [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+SELECT seriesPeriodDetectFFT([10,20,30,10,20,30,10,20,30, 10,20,30,10,20,30,10,20,30,10,20,30]);
+SELECT seriesPeriodDetectFFT([10.1, 20.45, 40.34, 10.1, 20.45, 40.34,10.1, 20.45, 40.34,10.1, 20.45, 40.34,10.1, 20.45, 40.34,10.1, 20.45, 40.34,10.1, 20.45, 40.34, 10.1, 20.45, 40.34]);
+SELECT seriesPeriodDetectFFT([10.1, 10, 400, 10.1, 10, 400, 10.1, 10, 400,10.1, 10, 400,10.1, 10, 400,10.1, 10, 400,10.1, 10, 400,10.1, 10, 400]);
+SELECT seriesPeriodDetectFFT([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]);
+SELECT seriesPeriodDetectFFT(arrayMap(x -> sin(x / 10), range(1000)));
+SELECT seriesPeriodDetectFFT(arrayMap(x -> abs((x % 6) - 3), range(1000)));
+SELECT seriesPeriodDetectFFT(arrayMap(x -> if((x % 6) < 3, 3, 0), range(1000)));
+SELECT seriesPeriodDetectFFT([1,2,3]);
+SELECT seriesPeriodDetectFFT(a) FROM tb1;
```
Comment (jkartseva) [inline tests/queries/0_stateless/02813_series_period_detect.sql:17]: And after:
```
DROP TABLE IF EXISTS tb1;
```

File: tests/queries/0_stateless/02813_series_period_detect.sql [hunk 9]
```diff
@@ -0,0 +1,21 @@
+-- Tags: no-fasttest
+
+drop table if exists tb1;
+
+create table tb1 (n UInt32, a Array(Int32)) engine=Memory;
+insert into tb1 values (1, [10,20,30,10,20,30,10,20,30, 10,20,30,10,20,30,10,20,30,10,20,30]), (2, [2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]), (3, [6,3,4]);
+
+SELECT seriesPeriodDetectFFT([139, 87, 110, 68, 54, 50, 51, 53, 133, 86, 141, 97, 156, 94, 149, 95, 140, 77, 61, 50, 54, 47, 133, 72, 152, 94, 148, 105, 162, 101, 160, 87, 63, 53, 55, 54, 151, 103, 189, 108, 183, 113, 175, 113, 178, 90, 71, 62, 62, 65, 165, 109, 181, 115, 182, 121, 178, 114, 170]);
```
Comment (jkartseva) [inline tests/queries/0_stateless/02813_series_period_detect.sql:8]: nit: Let's use either uppercase or lowercase SQL commands within a single *.sql file.

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+                    ErrorCodes::ILLEGAL_COLUMN,
+                    "Illegal column {} of first argument of function {}",
+                    arguments[0].column->getName(),
+                    getName());
+        }
+        return res;
+    }
+
+    template <typename T>
+    bool executeNumber(const IColumn & src_data, Float64 & period, ColumnArray::Offset &start, ColumnArray::Offset &end) const
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetectFFT.cpp:107]: Also, style:
`ColumnArray::Offset & start, ColumnArray::Offset & end)`

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+    {
+        const ColumnVector<T> * src_data_concrete = checkAndGetColumn<ColumnVector<T>>(&src_data);
+        if (!src_data_concrete)
+            return false;
+
+        const PaddedPODArray<T> & src_vec = src_data_concrete->getData();
+
+        size_t len = end - start;
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetectFFT.cpp:115]: Please add `chassert(start <= end)` before the subtraction.

File: docs/en/sql-reference/functions/time-series-functions.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+---
+
+# Time Series Functions
+
+Below functions are used for time series analysis.
+
+## seriesPeriodDetectFFT
+
+Finds the period of the given time series data using FFT
+Detect Period in time series data using FFT.
```
Comment (jkartseva) [inline docs/en/sql-reference/functions/time-series-functions.md:14]: Lines 13 and 14 essentially tell the same. Also, a dot is missing at the end of the sentence.

File: docs/en/sql-reference/functions/time-series-functions.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+FFT - Fast Fourier transform (https://en.wikipedia.org/wiki/Fast_Fourier_transform)
```
Comment (jkartseva) [inline docs/en/sql-reference/functions/time-series-functions.md:15]: The link can be formatted:
`[Fast Fourier transform](https://en.wikipedia.org/wiki/Fast_Fourier_transform)`

**Review by jkartseva — APPROVED** [review]
LGTM with a few nits and suggestions.

**Pull request merged**